File: 0081-search-in-rotated-sorted-array-ii/0081-search-in-rotated-sorted-array-ii.java
class Solution {
    public boolean search(int[] nums, int target) {
        if(nums == null || nums.length == 0){
            return false;
        }
        int low  = 0;
        int high  = nums.length - 1;

        while(low <= high){
            int mid = low  + (high - low) /2;

            if(nums[mid] ==  target){
                return true;
            }
            //duplicates;
            if(nums[low] == nums[mid] && nums[mid] == nums[high]){
                low++;
                high--;
                continue;
            }
            //find sorted half
            if(nums[low] <= nums[mid]){
                if(nums[low] <= target && target <= nums[mid]){
                    high = mid -1;
                } else {
                    low = mid +1;
                }

            } else {
                if(nums[mid] <= target && target <= nums[high]){
                    low = mid + 1;
                } else {
                    high = mid -1;
                }
            }
        }

        return false;
    }
}

File: 0081-search-in-rotated-sorted-array-ii/0081-search-in-rotated-sorted-array-ii.py
class Solution(object):
    def search(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: bool
        """
        if not nums:
            return False
        
        low, high = 0, len(nums) - 1

        while low <= high:
            mid = low + (high - low) // 2

            if nums[mid] == target:
                return True

            if nums[low] == nums[mid] == nums[high]:
                low += 1
                high -= 1

            elif nums[low] <= nums[mid]:
                if nums[low] <= target < nums[mid]:  
                    high = mid - 1  
                else:
                    low = mid + 1  
                                
            else:
                if nums[mid] < target <= nums[high]:  
                    low = mid + 1  
                else:
                    high = mid - 1  

        return False
